db/types/inference.py

```diff
@@ -0,0 +1,77 @@
+import logging
+from sqlalchemy import MetaData, Table
+from sqlalchemy import VARCHAR, TEXT, Text
+from db.types import alteration
+
+logging.basicConfig(
+    format='%(asctime)s - %(name)s - %(levelname)s:  %(message)s',
+    level=logging.INFO
+)
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+TYPE_INFERENCE_DAG = {
+    alteration.BOOLEAN: [],
+    alteration.EMAIL: [],
+    alteration.INTERVAL: [],
+    alteration.NUMERIC: [
+        alteration.BOOLEAN,
+    ],
+    alteration.STRING: [
+        alteration.BOOLEAN,
+        alteration.NUMERIC,
+        alteration.INTERVAL,
+        alteration.EMAIL,
+    ],
+}
+
+
+class DagCycleError(Exception):
+    pass
+
+
+def infer_column_type(
+        schema,
+        table_name,
+        column_name,
+        engine,
+        depth=0,
+        type_inference_dag=TYPE_INFERENCE_DAG,
+):
+    if depth > 100:
+        raise DagCycleError("The type_inference_dag likely has a cycle")
+    supported_types = alteration.get_supported_alter_column_types(engine)
+    reverse_type_map = {
+        Text: alteration.STRING,
+        TEXT: alteration.STRING,
+        VARCHAR: alteration.STRING,
+    }
+    reverse_type_map.update({v: k for k, v in supported_types.items()})
+    with engine.begin():
+        metadata = MetaData(bind=engine, schema=schema)
+        column_type = Table(
+            table_name, metadata, schema=schema, autoload_with=engine,
+        ).columns[column_name].type.__class__
+    column_type_str = reverse_type_map.get(column_type)
+    logger.debug(f"column_type_str: {column_type_str}")
+    for type_str in type_inference_dag.get(column_type_str, []):
+        try:
+            alteration.alter_column_type(
+                schema, table_name, column_name, type_str, engine,
+            )
+            logger.info(f"Column {column_name} altered to type {type_str}")
+            column_type = infer_column_type(
+                schema,
+                table_name,
+                column_name,
+                engine,
+                depth=depth + 1,
+                type_inference_dag=type_inference_dag,
+            )
+            break
+        except Exception:
+            logger.info(
```

Comment:
Shouldn't this be `logger.error`?

Reply:
This exception is an expected part of the control flow (it's how we know when a given type is inappropriate for a given column).  I think `info` is reasonable for that.  I _do_ think I should remove the spot where I set the log level to `INFO` in this file, though (it should be set by the importer if they want a non-default level).  What do you think?

Reply:
Can you catch the specific exception you're expecting and add a comment explaining it's the expected flow?

I agree re: removing log level in the file.

Reply:
Unfortunately, there's a bunch of different exceptions based on which type was casting to which other.  I can try to figure out if they're sub-classes of something appropriate (I think python will catch the parent class in that case, I'll have to check)

Reply:
Okay, we'll use `sqlalchemy.exc.DatabaseError`.  It's not super-precise, but it at least restricts to "you asked to do something and the DB said no".